### PR TITLE
Change oaDOI to Unpaywall

### PR DESCRIPTION
**Pull request by bondjimbond**
**JIRA Ticket**: (https://jira.duraspace.org/browse/ISLANDORA-2224)

# What does this Pull Request do?

Removes the oaDOI badge, creates an Unpaywall badge to replace it.

# What's new?
New name, new API endpoint, adjustment to data processing, new admin form fields (email, API endpoint)

# How should this be tested?

- Have an object with a DOI that is recognized by Unpaywall (easy if you've got oaDOI installed)
- Check out this branch, enable module, enable block
- See if block appears on that object
- Make sure block doesn't do anything else that's weird

# Additional Notes:
This may need an update hook to uninstall oaDOI...? I don't know how that works.

# Interested parties
@rosiel  @Islandora/7-x-1-x-committers

### CODE REVIEW

**Comment by DiegoPino** [comment]
Eh... this won't work. You are deprecating a Module, even if its a submodule, it is a module. nop. won't work. Want to discuss this on the next call? @rosiel @bryjbrown ?

**Review by DiegoPino — DISMISSED** [review]
Nope. deprecating old modules does not work this way. sorry

**Comment by bondjimbond** [comment]
OK, let me know how it does work and I'll fix

**Comment by DiegoPino** [comment]
@bondjimbond you would have to keep the old one for a time, give it a deprecation notice and create your new one in parallel, which I guess is the same really so does not need Licence agreement process really, but requires committers agreement because it will require manual work for people with running Scholar based repositories.

**Comment by bondjimbond** [comment]
@DiegoPino I think you left the meeting before we talked about this ticket, but we did discuss at the committers' call today. Deprecating old module this release, and adding the replacement in 7.x and 7.x-1.12... unless I missed something, @rosiel?

**Comment by DiegoPino** [comment]
Yeah, sadly, I had to leave. Actually, that idea of deprecating was mine when dealing with that blocker JIRA, in a talk with @rosiel on Tuesday. But deprecating is different than removing.

**Comment by DiegoPino** [comment]
@bondjimbond reading through Unpaywall requirements and limits (100.000 a day which not sure how  this module handles once reached), can we add a link to http://unpaywall.org/products/api?  in the README? Also, can you share a DOI that works for testing?

We need (also) to add a deprecation notice for the old module:

- I would suggest a change in the old's module README,
- A date when this will happen (next release, did you discuss this yesterday?)

And maybe a PHP generated warning/message.. the tone, the type and the where can vary...
```PHP
$message = format_string('The Module @modulename as of 7.x-1.11 has been deprecated in favour of  @othermodulename .', array(
          '@modulename' => 'oldie'
          '@othermodulename' => 'newbie',
        ));
 trigger_error(filter_xss($message), E_USER_DEPRECATED);
````
More at http://php.net/manual/en/function.trigger-error.php

@adam-vessey summoning your wisdom here regarding the message, the where and the tone.  Do we have any previous example of something like this? This is not the same as deprecating a top-level module. It is a bit tricky since the update procedure for `badges` is a `git pull` or a full download of the module (parent to this), which, once this is deprecated and removed in next release, will make "Drupal uncomfortable" because it will remove the old submodule without warning by just deleting the folder...

Lastly any settings we can/need to move from one module to another `automagically`?

 Pinging @rosiel since she led the deprecation conversation during the committers call yesterday

**Comment by bondjimbond** [comment]
Working DOI: 10.3758/BRM.40.4.1075

Link to API info in readme: done

Deprecation notice: to come; will wait for response from @adam-vessey to @DiegoPino's questions.

Settings to move: There's only one setting in islandora_oadoi: islandora_oadoi_linktext. It would map to islandora_unpaywall_linktext. How would migrating that work?

**Comment by DiegoPino** [comment]
> Settings to move: There's only one setting in islandora_oadoi: islandora_oadoi_linktext. It would map to islandora_unpaywall_linktext. How would migrating that work?

I guess, something like 

- in one of the  .install functions, 

  - if [old module is installed](https://api.drupal.org/api/drupal/includes%21module.inc/function/module_exists/7.x)

  - [get variable](https://api.drupal.org/api/drupal/includes%21bootstrap.inc/function/variable_get/7.x) value 
   -  and set new value using [that one](https://api.drupal.org/api/drupal/includes%21bootstrap.inc/function/variable_set/7.x)?

**Comment by adam-vessey** [comment]
@DiegoPino @bondjimbond: Yeah, I believe this is new ground, deprecating a submodule... I'm not sure we could really automatically do the full setup of the new submodule, as it requires additional parameters, IIRC?

It might be safest to pare all the functionality out of the existing submodule, leaving just a `.info`, `.module` and `.install` (and README and LICENSE?... just a skeleton, essentially), and throw an implementation of `hook_update_N()` in the deprecated/obsolete module to inform admins that the module is deprecated/obsoleted due to whatever upstream API change... If we had sufficient info, we could [enable](https://api.drupal.org/api/drupal/includes%21module.inc/function/module_enable/7.x) and configure the new module at this point, and possibly delete the variables for the module, as they'd no longer serve a purpose... on second thought, if the additional parameter is just an e-mail address, we could possibly default to grabbing the system/site e-mail address?... unsure if this would be desirable... in any case, leaving the module in place with no functionality in the interim, pending its deletion should avoid "distressing" Drupal.

Seems... cleaner in my mind to have the pared down module push its configuration to the new module instead of having the new place try to pull it from the old, given the old module will go away eventually... if it's being pulled in the new module, then we'd eventually be stuck with having to deal with references to variables originally declared in old modules which seems... messy... not the end of the world, but some complexity we could avoid foisting on our future selves.

Something of an unlikely issue, but... if there are any modules in the wild which depend on calling and using any of the hook responses that are implemented, there could be issues... if they call the function directly, like the entirely hypothetical:

```php
function example_block_view($delta = '') {
  if ($delta == 'example_block') {
    $block = islandora_oadoi_block_view('islandora_oadoi_badge');
    // Alter $block in some way...
    return $block;
  }
}
```

... such an integration would be broken by nuking things... if the new module is essentially just renaming... the old module could be made the throw the deprecation message everywhere its functions are effectively used, have an update hook enabling the new module, and `module_exists()` checks before calling out to the new functions... dunno...

... how far do we want to go?

**Comment by bondjimbond** [comment]
@adam-vessey I like the idea of putting the update stuff into the old module rather than the new one, if it works. 

> I'm not sure we could really automatically do the full setup of the new submodule, as it requires additional parameters, IIRC?

The API Endpoint variable shouldn't change from the default, so that one is safe to leave alone. The email address -- I don't think we should make assumptions about it. Could we just add a note in the deprecation notice that the new module requires an email address to be configured?

> Something of an unlikely issue, but... if there are any modules in the wild which depend on calling and using any of the hook responses that are implemented, there could be issues...

We can't control for people's custom work. A deprecation notice should be enough to alert them that they need to change their custom code, no?

**Comment by DiegoPino** [comment]
> We can't control for people's custom work. A deprecation notice should be enough to alert them that they need to change their custom code, no?

I agree on that.

Also, can we "not partly nuke", just notify folks via notices and install hooks and put a deadline and then totally nuke? means, just keep all around until de deprecate/deprecate, by just adding, for now, notices and alerts all around? I feel removing pieces-stripping down the old module we will just introduce more issues but maybe I'm missing some benefits.

+1 on "putting the update stuff into the old module rather than the new one, if it works." too. In general "if it works" conditions my support  😄

**Comment by bondjimbond** [comment]
Latest commit: Readme updated, deprecation notice added in block_view function.

I'm going to need some guidance on where and how to add the module_enable and variable passing functions in islandora_oadoi. @adam-vessey might you be willing to add a PR to my branch if you have the time and know offhand how it should be done? If not, some general guidance would be appreciated, and I'll try to handle it myself.

**Comment by bondjimbond** [comment]
@DiegoPino @adam-vessey Any ideas on how and where to add the module_enable and variable-passing functions to islandora_oadoi?

**Comment by adam-vessey** [comment]
@bondjimbond: No strong feeling any which way... where do you want to have the new one enabled and configured?

**Comment by bondjimbond** [comment]
@adam-vessey I'd like to do as discussed, and have code in islandora_oadoi execute during update.php to the effect of the following pseudocode:

```
 IF (islandora_oadoi is enabled) {
    module_enable(islandora_unpaywall);
    islandora_unpaywall_link_text = islandora_oadoi_link_text;
 }
```

No idea how to actually make that happen, though. Would you be willing to make a pull against my branch to add that functionality?

**Comment by adam-vessey** [comment]
@bondjimbond: I don't have the cycles, unfortunately. I recommend taking a look at the hooks available... in particular, the [aforementioned](https://github.com/Islandora/islandora_badges/pull/19#issuecomment-390272073) [`hook_update_N()`](https://api.drupal.org/api/drupal/modules!system!system.api.php/function/hook_update_N/7.x).

@DiegoPino's [mentioned workflow](https://github.com/Islandora/islandora_badges/pull/19#issuecomment-390252950) is pretty much it, no?

**Comment by DiegoPino** [comment]
@bondjimbond I could (later tonight) make that pull against yours but then @adam-vessey would have to review and merge if you all are OK with that. I can also guide you through that dark forest (really not that dark, mostly 2-3 trees and a few kids in troll costumes)

**Comment by bondjimbond** [comment]
Thanks, @DiegoPino - if you've got the time and inclination, I'd love some guidance. I'm trying out the following in islandora_oadoi.install, but it isn't doing anything:

```
function islandora_oadoi_update () {
  if (module_exists('islandora_unpaywall')) {
    module_enable(array('islandora_unpaywall'));
    $oadoi_variable = variable_get('islandoar_oadoi_linktext', 'Link to PDF');
    variable_set('islandora_unpaywall_linktext', $oadoi_variable);
  }
}
```

**Comment by DiegoPino** [comment]
Hi, you just need to rename that hook to islandora_oadoi_update_7100
See https://api.drupal.org/api/drupal/modules%21system%21system.api.php/function/hook_update_N/7.x
for the semantics behing the different numbers

**Comment by bondjimbond** [comment]
@DiegoPino Thanks! That got update.php to recognize it. But it didn't enable islandora_unpaywall.

`    module_enable(array('islandora_unpaywall'));`

Am I missing something?

**Comment by DiegoPino** [comment]
Hu, I did not know you could enable another module via a hook? Let me research this!

**Comment by bondjimbond** [comment]
It also is not passing the value of the islandora_oadoi_linktext to islandora_unpaywall_linktext.

```
    $oadoi_variable = variable_get('islandora_oadoi_linktext', 'Link to PDF');
    variable_set('islandora_unpaywall_linktext', $oadoi_variable);
```

**Comment by bondjimbond** [comment]
@DiegoPino @adam-vessey OK, so here's the problem...

`module_exists` only returns TRUE if the module is currently enabled. So if it is not already enabled, well, nothing happens. If it is already enabled, then there's really nothing to do - the user has probably already configured islandora_unpaywall.

Should this update hook be discarded? Or perhaps if islandora_unpaywall exists we should disable islandora_oadoi?

**Comment by DiegoPino** [comment]
The later @bondjimbond because you wrote islan**doar**_oadoi_linktext instead of islandora_oadoi_linktext?

**Comment by DiegoPino** [comment]
"the user has probably already configured islandora_unpaywall." that is a statement that assumes maybe too much?

**Comment by bondjimbond** [comment]
> The later @bondjimbond because you wrote islandoar_oadoi_linktext instead of islandora_oadoi_linktext?

I corrected that in my local copy before I complained that it doesn't work. :)  Even with correct spelling it doesn't work. But I think that's perhaps because the module needs to be enabled for module_exists() to pick it up?

> "the user has probably already configured islandora_unpaywall." that is a statement that assumes maybe too much?

Possibly? But then we're also assuming that the user wants the islandora_oadoi linktext to be pushed to islandora_unpaywall after they've already enabled islandora_unpaywall. Hard to know what people want.

It's a new module -- maybe it's safe to just say "New module, new configuration". Especially considering that (a) they still need to put in their email address, and (b) they will have to go back to Structure -> Blocks and put the new block in the old islandora_oadoi place.

**Comment by bondjimbond** [comment]
@DiegoPino @adam-vessey 

Since it looks like the variable_set option won't work, I've changed the update hook to:

1. Disable and uninstall islandora_oadoi
2. Tell user that islandora_oadoi is deprecated and instruct them to install, enable, and configure islandora_unpaywall.

Will that do the trick? Or should I skip the uninstall line and tell them to uninstall manually so users can still refer to existing configuration when setting up the new module?

File: modules/islandora_oadoi/islandora_oadoi.install
Comment (whikloj) [inline modules/islandora_oadoi/islandora_oadoi.install]: I don't think this is needed as for this code to be run the `islandora_oadoi` module would have to be installed.

Comment (DiegoPino) [reply modules/islandora_oadoi/islandora_oadoi.install]: Not sure why that change. if (module_exists('islandora_unpaywall')) should be the right now,  means checking for new the module to "migrate" the setting from old to new? If this becomes too complicated (does not seem to but who knows?) we should just document that people need to use their mouse and keyboard to migrate the value there or something like that

File: modules/islandora_unpaywall/islandora_unpaywall.install
Comment (whikloj) [inline modules/islandora_unpaywall/islandora_unpaywall.install]: I saw some discussion of this on the conversation page from @adam-vessey. You could add an `islandora_unpaywall_install()` that checks `if (module_exists("islandora_oadoi")) {` and copies the `islandora_oadoi_linktext` variable to `islandora_unpaywall_linktext`.

Comment (DiegoPino) [reply modules/islandora_unpaywall/islandora_unpaywall.install]: It seems there was an agreement on doing this on the old module https://github.com/Islandora/islandora_badges/pull/19#issuecomment-390300454 But i agree with Jared that having this in the new one could solve the issue. After a few releases (like islandora 7.x-1.20 😨 ) we could then remove the code

File: modules/islandora_oadoi/islandora_oadoi.install
Comment (bondjimbond) [reply modules/islandora_oadoi/islandora_oadoi.install]: It turns out that `module_exists()` only checks for whether the module is enabled, not whether it is in the system. So this code checks if islandora_oadoi is enabled, and if so, disables it (because it doesn't work anymore).

@DiegoPino The original concept was to check if islandora_unpaywall is present and then enable it if it is, but `module_exists()` doesn't actually do that. So I removed that piece.

File: modules/islandora_unpaywall/islandora_unpaywall.install
Comment (whikloj) [reply modules/islandora_unpaywall/islandora_unpaywall.install]: Yeah this is a personal preference type thing, I thought the `install` was easier because if your tell people to install the new module, then uninstall the old one you don't have to run any update hooks to get the variable copied. Really people could also just be told to check their settings again.

Comment (bondjimbond) [reply modules/islandora_unpaywall/islandora_unpaywall.install]: Would `islandora_unpaywall_install()` only trigger at the point of first installing the module, or every time it's enabled?

File: modules/islandora_oadoi/islandora_oadoi.install
Comment (whikloj) [reply modules/islandora_oadoi/islandora_oadoi.install]: @bondjimbond but would this update hook get run if `islandora_oadoi` is not enabled? I can't see that it would, in which case I'm not sure this check is achieving anything.

File: modules/islandora_unpaywall/islandora_unpaywall.install
Comment (whikloj) [reply modules/islandora_unpaywall/islandora_unpaywall.install]: First install only, unless they uninstall (not disable) and re-install this module.

File: modules/islandora_oadoi/islandora_oadoi.install
Comment (bondjimbond) [reply modules/islandora_oadoi/islandora_oadoi.install]: @whikloj Just tested - yes, it does get run even if the module is not enabled.

File: modules/islandora_unpaywall/islandora_unpaywall.install
Comment (whikloj) [reply modules/islandora_unpaywall/islandora_unpaywall.install]: Again, the procedure is 
1. Install `islandora_unpaywall`
1. Uninstall `islandora_oadoi`

Then this would not be a concern as any subsequent installs of `islandora_unpaywall` would not have an `islandora_oadoi` to copy a variable from.

File: modules/islandora_oadoi/islandora_oadoi.install
Comment (whikloj) [reply modules/islandora_oadoi/islandora_oadoi.install]: @bondjimbond that seems wrong according to the [drupal documentation](https://api.drupal.org/api/drupal/includes%21module.inc/function/module_exists/7.x). But ok.

File: modules/islandora_unpaywall/islandora_unpaywall.install
Comment (bondjimbond) [reply modules/islandora_unpaywall/islandora_unpaywall.install]: So I should put `drupal_uninstall_modules(array('islandora_oadoi'));` in `islandora_unpaywall.install`, after `islandora_unpaywall_install()` copies the variable?

File: modules/islandora_oadoi/islandora_oadoi.install
Comment (whikloj) [reply modules/islandora_oadoi/islandora_oadoi.install]: Oh wait, I see my mistake. Nevermind 🤦‍♂️

File: modules/islandora_unpaywall/islandora_unpaywall.install
Comment (whikloj) [reply modules/islandora_unpaywall/islandora_unpaywall.install]: You can, just be sure to output information to the screen that you are doing that. People can be confused when one module uninstalls another one 😄

Comment (bondjimbond) [reply modules/islandora_unpaywall/islandora_unpaywall.install]: Message added with latest commit.

**Comment by bondjimbond** [comment]
OK, finally passing. Travis is very picky about drupal_set_message in hook_install. There were also some small errors, but it seems to be working properly now.

- Deprecation notice in update hook
- Enabling/installing islandora_unpaywall automatically pulls in the linktext from islandora_oadoi, disables and uninstalls islandora_oadoi
- Added a warning message to block_view if email address is not configured
- Finally appeased Travis

To test:
- Uninstall islandora_unpaywall if installed
- Enable islandora_oadoi, add to Sidebar Second block region
- View an object with unpaywall-approved DOI (10.3758/BRM.40.4.1075), observe deprecation notice
- Configure custom linktext in islandora_oadoi
- Run update.php to view deprecation message
- Enable islandora_unpaywall, observe notice
- Add islandora_unpaywall to Sidebar Second block region
- Check that islandora_oadoi is now disabled and uninstalled
- Check that custom linktext is in islandora_unpaywall config
- View an object with unpaywall-approved DOI (10.3758/BRM.40.4.1075)
- Observe no badge, warning message to configure email address
- Observe working Unpaywall badge

@adam-vessey @DiegoPino @whikloj Anything left to do, or does this cover it all?

**Comment by rosiel** [comment]
A few suggestions (not blocking, but the kind of "minor annoyances" that might make this better/easier to do right if you're not following this PR word-for-word):
- in the islandora_oadoi.info file, change the display name to "Islandora oaDOI (Deprecated)" and the description to "Islandora oaDOI linking - Deprecated in favour of Islandora Unpaywall (islandora_unpaywall)." That'll reduce the likelihood of someone turning it on *after* running this update. 
- in the islandora_unpaywall.info file, add the configuration link as described here: https://www.drupal.org/docs/7/creating-custom-modules/writing-module-info-files-drupal-7x. The config link can be hard to find, especially if you don't have admin_menu installed. And it'll be right there when you turn the module on! 
- When rendering the Unpaywall block, the module will complain if the email is the default (or blank). Could the admin form do the same in a validate function (not making the form invalid, just providing some warning to the hapless admin who might not have a citation with a DOI ready to test. Right now, it looks like your configuration is a-ok with the default, but an ugly warning will show up (to everyone) on objects with DOIs. It's this block:
```php
if ($email == 'email@example.com' || $email == '') {
    drupal_set_message(t('The Unpaywall API requires a valid email address. Set one in the Islandora Unpaywall configuration.'), 'warning');
}
```

**Comment by bondjimbond** [comment]
Thanks @rosiel - all your points make sense and have been added to latest commit.

@DiegoPino Is everything you were concerned about addressed? Are you ready to approve?

**Review by rosiel — APPROVED** [review]

**Comment by rosiel** [comment]
Squashed and merged - @bondjimbond can you cherrypick it into a 7.x-1.11 PR?

**Comment by bondjimbond** [comment]
@rosiel Not sure about cherry-picking, but I'll make a new PR for it.

**Pull request merged**
